src/v/cloud_storage/remote_partition.cc

```diff
@@ -786,79 +957,62 @@ ss::future<> remote_partition::erase(ss::abort_source& as) {
-    } else if (manifest_get_result == download_result::failed) {
-        // Treat non-timeout errors as permanent, to avoid controller
-        // getting stuck.
-        vlog(
-          _ctxlog.warn,
-          "Error downloading manifest: objects will not be deleted for "
-          "this "
-          "topic");
-        co_return;
-    }
-
     // Having handled errors above, now our partition manifest fetch was
     // either a notfound (skip straight to erasing topic manifest), or a
     // success (iterate through manifest deleting segements)
-    if (manifest_get_result != download_result::notfound) {
-        // Erase all segments
-        for (const auto& i : manifest) {
-            auto path = manifest.generate_segment_path(i.second);
-            vlog(_ctxlog.debug, "Erasing segment {}", path);
-            // On failure, we throw: this should cause controller to retry
-            // the topic deletion operation that called us, until it
-            // eventually succeeds.
-            // TODO: S3 API has a plural delete API, which would be more
-            // suitable.
-            if (co_await tolerant_delete_object(
-                  _bucket,
-                  cloud_storage_clients::object_key(path),
-                  local_rtc)) {
-                co_return;
-            };
-
-            auto tx_range_manifest_path
-              = tx_range_manifest(path).get_manifest_path();
-            if (co_await tolerant_delete_object(
-                  _bucket,
-                  cloud_storage_clients::object_key(tx_range_manifest_path),
-                  local_rtc)) {
-                co_return;
-            };
+
+    // Erase all segments
+    const size_t batch_size = 1000;
+    auto segment_i = manifest.begin();
+    while (segment_i != manifest.end()) {
+        std::vector<cloud_storage_clients::object_key> batch_keys;
+        batch_keys.reserve(batch_size);
+
+        std::vector<cloud_storage_clients::object_key> tx_batch_keys;
+        tx_batch_keys.reserve(batch_size);
+
+        for (size_t k = 0; k < batch_size && segment_i != manifest.end(); ++k) {
+            auto segment_path = manifest.generate_segment_path(
+              segment_i->second);
+            batch_keys.push_back(
+              cloud_storage_clients::object_key(segment_path));
+            tx_batch_keys.push_back(cloud_storage_clients::object_key(
+              tx_range_manifest(segment_path).get_manifest_path()));
+            segment_i++;
         }
 
-        // Erase the partition manifest
-        vlog(_ctxlog.debug, "Erasing partition manifest {}", manifest_path);
-        if (co_await tolerant_delete_object(
-              _bucket,
-              cloud_storage_clients::object_key(manifest_path),
-              local_rtc)) {
+        vlog(
+          _ctxlog.debug,
+          "Erasing segments {}-{}",
+          *(batch_keys.begin()),
+          *(--batch_keys.end()));
+
+        if (co_await tolerant_delete_objects(
```

Comment:
@jcsp looks like `manifest` can be a reference to the class member `_manifest`. in that case the iteration combined with co_await is a dark pattern worth asking about: can the iterators be invalidated by asynchronous changes to `_manifest` member?

Reply:
At this point in the lifeycle, raft has been shut down, so there shouldn't be anything (like stm replay) touching manifest in the background.  I'll bear this in mind though, it would be nice to avoid relying on that kind of implicit knowledge.